File: lab 7/java/src/Simulation.java
import java.util.LinkedList;

public class Simulation
{
    LinkedList<Philosopher> philosophers;

    Simulation(LinkedList<Philosopher> philosophers)
    {
        this.philosophers = philosophers;
    }

    long average = 0;
    long max = 0;
    long min = Long.MAX_VALUE;

    public String simulate()
    {
        String result = "";
        try
        {
            for(Philosopher philosopher: philosophers)
            {
                philosopher.start();
            }
            for(Philosopher philosopher: philosophers)
            {
                philosopher.join();
            }
            for(Philosopher philosopher: philosophers)
            {
                long time = philosopher.waitingTime/philosopher.waitingCount;
                average = average + time / philosophers.size();
                max = Math.max(max, time);
                min = Math.min(min, time);
            }
            result = average + "";
        }
        catch(InterruptedException e)
        {
            e.printStackTrace();
        }
        return result;
    }
}

File: lab 7/scala/src/main/scala/Simulation.scala
import scala.collection.mutable.ListBuffer


class Simulation(var philosophers: ListBuffer[Philosopher]) {
  var average = 0
  var max = 0
  var min = Int.MaxValue

  def simulate: String = {
    var result = ""
    try {
      for (philosopher <- philosophers) {
        philosopher.start()
      }
      for (philosopher <- philosophers) {
        philosopher.join()
      }
      for (philosopher <- philosophers) {
        val time = philosopher.waitingTime / philosopher.waitingCount
        average = average + time / philosophers.size
        max = Math.max(max, time)
        min = Math.min(min, time)
      }
      result = " " + average
    } catch {
      case e: InterruptedException =>
        e.printStackTrace()
    }
    result
  }
}
